Add module for obtaining load statistics from cgroupstats

alexdias opened the pull request:
Hi guys,

This pull request adds a new module (very akin to taskstats.py) that allows load statistics to be obtained from cgroupstats. There's also an example script, which can be used like such:
```
$ sudo python2.7 examples/cgroupstats_dump.py /sys/fs/cgroup/cpu/
LoadStats(nr_sleeping=88, nr_running=0, nr_stopped=0,nr_uninterruptible=0, nr_iowait=0)
```

Let me know what you think!

vmauge commented:
Before doing a longer review can you please fix the following point :
- Run flake8 against your diff to fix ident and other code formating issue

In addition, can you confirm if you make the code as a individual contributor or part of your company ?
- If it's the first please update reassign copyright to Facebook. As part of the CLA agreement (https://code.facebook.com/cla), you agreed to give use the copyright. 
- If it's part of company can you confirm which company so I can check if there is corporate CLA approved (https://developers.facebook.com/opensource/cla?type=company)

alexdias commented:
@vmauge Hi Vincent,

Regarding the first point, I did and the failures were on pre-existing code; I chose not to fix them but I can if you'd like me to? Here's the output of flake8:

```
Alexandres-MacBook-Pro:gnlpy alex$ flake8 .
./netlink.py:148:1: E305 expected 2 blank lines after class or function definition, found 1
./netlink.py:353:1: E305 expected 2 blank lines after class or function definition, found 1
./netlink.py:405:1: E305 expected 2 blank lines after class or function definition, found 1
./taskstats.py:60:1: E305 expected 2 blank lines after class or function definition, found 1
./examples/ipvs_dump.py:64:1: E305 expected 2 blank lines after class or function definition, found 1
./examples/taskstats_dump.py:28:1: E305 expected 2 blank lines after class or function definition, found 1
./tests/ipvs_tests.py:560:1: E305 expected 2 blank lines after class or function definition, found 1
```

This is the same as the output given on the travis CI build result.

Did you see any more code formatting issues (in my diff in particular) besides the ones just above?

I'll get back to you tomorrow regarding your second point.

chantra commented:
ok, so the flake8 issue is most likely due to flake8 version (and its dependencies) change: flake8-2.5.4 vs flake8-3.3.0.
I will fix this in a separate diff so you can rebase your diff on a clean master.

chantra commented:
#20 will address the lint errors.

chantra commented on cgroupstats.py:
```diff
@@ -0,0 +1,77 @@
+
+    def __init__(self, verbose=False):
+        self.verbose = verbose
+        self.nlsock = netlink.NetlinkSocket()
```
pass verbose to the `NetlinkSocket`. Honestly, this should have been abstracted into a parent class and commonly used by `ipvs`, `taskstats` and not `CgroupstatsClient`

alexdias replied:
@chantra Ah, thanks for the catch! How do you mean in terms of abstracting it? The usage of `NetlinkSocket` ?

chantra replied:
I mean that all clients are inherently NetlinkClient, which will use the same `__init__` of extend it. This is more of a thought and not a call for action on your side.

chantra commented on cgroupstats.py:
```diff
@@ -0,0 +1,77 @@
+    'CgroupstatsMessage', 'TASKSTATS',
+    ('NULL', netlink.IgnoreType),  # Reserved
+    ('GET', netlink.IgnoreType),
+    ('NEW', netlink.IgnoreType),
```
IIRC, the kernel will emit a `NEW` event back, and this what is going to be used to trigger the decapsulation. So I think, what we would want here is:
```
     ('NEW', CgroupstatsType),
```

where `CgroupstatsType` is similar to `TaskstatsType`, which is made of `netlink.U64Type` type. With each fields corresponding to their equivalent in `LoadStats`, e.g `NR_SLEEPING`, `NR_RUNNING`....

`LoadStats` will become unnecessary as there is no need for complex types.

chantra replied:
http://lxr.free-electrons.com/source/kernel/taskstats.c#L410 for kernel code.

alexdias replied:
@chantra What I've done is a bit of a hack, and it works because the `CGROUPSTATS_CMD_MAX` field ends up being equivalent to `4`; `STATS` is equivalent to `5`. This is what the kernel expects (in terms of integers), because as you can see here http://lxr.free-electrons.com/source/include/uapi/linux/cgroupstats.h the enum is initialised to `__TASKSTATS_CMD_MAX`, which is `3` (seen here: http://lxr.free-electrons.com/source/include/uapi/linux/taskstats.h).

So in order to only have `GET` and `NEW` (which I agree that we should do here), and therefore have them be equivalent to `4` and `5`, I think we need to change netlink.py (if I'm not mistaken). Basically the issue is that here https://github.com/facebook/gnlpy/blob/master/netlink.py#L301 the key is is taken from the attribute's position in the fields, and there isn't a way to initialise the key to a value.

As for the complex type, as far as I could see the kernel emits an attribute with the following structure:
[AttributeLength (4 bytes), NrSleeping (8 bytes), NrRunning (8 bytes), NrStopped (8 bytes), NrUninterruptible (8 bytes), NrIoWait (8 bytes)]

So if I define it as an attribute list type of 5 different fields of `netlink.U64Type` type each, the issue I have is that the code expects the kernel to send 5 different structures of [AttributeLength, Attribute], but from what I could see that doesn't match with what the kernel actually sends.

I could define it as one single attribute, with a custom type that would correspond to 40 bytes (5 * 8 bytes). But I would then still need to decapsulate the 5 different items onto some structure so they could be accessed separately.

Let me know if you want me to paste some example data that the kernel sends!

chantra replied:
For the complex type, you are totally right, leave it like this then. Let's call it `Cgroupstats` like in `include/uapi/linux/cgroupstats.h`

Also, you should not have to strip the first 4 bytes and rather let gnlpy do it. Those 4 bytes are made half of attr length, other half is attr type: https://github.com/facebook/gnlpy/blob/master/netlink.py#L35

This is where http://lxr.free-electrons.com/source/include/uapi/linux/cgroupstats.h#L57 comes into play.

So something along:

```python
CgroupstatsType =  netlink.create_attr_list_type(
    'CgroupstatsType',
    ('CGROUP_STATS', Cgroupstats)
```

Where `Cgroupstats` is your current `LoadStats`.

and then modify the message like:
```python
...
...
    ('GET', CgroupstatsCmdAttrList),
    ('NEW', CgroupstatsType),
    required_modules=[],
)
```

Does that make sense?

chantra commented on examples/cgroupstats_dump.py:
```diff
@@ -0,0 +1,33 @@
+
+    args = parser.parse_args(argv[1:])
+    c = CgroupstatsClient()
+    fd = os.open(args.path, os.O_RDONLY)
```
this also seems closer to the kernel api: http://lxr.free-electrons.com/source/tools/accounting/getdelays.c#L427

alexdias replied:
@chantra Ah yeah that seems better, thanks!

alexdias commented:
@chantra Thanks for the review!

I'll address your comments individually, while also adding a test and rebasing your style fixes.

alexdias commented on cgroupstats.py:
```diff
@@ -0,0 +1,77 @@
+import gnlpy.netlink as netlink
+
+
+class LoadStats(object):
```
@chantra Sure thing!

alexdias commented on cgroupstats.py:
```diff
@@ -0,0 +1,77 @@
+
+
+CgroupstatsMessage = netlink.create_genl_message_type(
+    'CgroupstatsMessage', 'TASKSTATS',
```
@chantra I believe the family name has to be TASKSTATS? If I change it to CGROUPSTATS, I get the following error:
`RuntimeError: ENOENT: No such file or directory. Extra info: (CtrlMessage(cmd=u'GETFAMILY', attr_list=CtrlAttrList(family_name=u'CGROUPSTATS'), version=0x1, flags=0x1), '')`

chantra commented on cgroupstats.py:
```diff
@@ -0,0 +1,77 @@
+    @staticmethod
+    def unpack(val):
+        fmt = 'QQQQQ'  # as per __fields__ above, and they're uint64 each
+        val = val[4:]  # strip attribute length, first 4 bytes
```
this can go based on my last comment and the use of `CgroupstatsType`

alexdias commented:
@chantra Sorry it's taken me so long to address this!

It's been super busy here in the past couple of weeks.

I've added some commits that addressed your comments. I still need to add tests.
I have a question regarding integration testing: we'll just be testing that we get some output from the kernel, right? We could do something like run a load-generating function (or process), and actually compare the output to an expected result, but it seems like that could be flakey.

chantra commented:
@alexdias yeah, I think the integration test will only need to confirm we can access the data without breaking. This would be enough to detect kernel api change, bug that get introduced... The actual result is pretty much the responsibility of the kernel.

alexdias commented:
@chantra I've just added a test, could I get another review please?

Thank you!

alexdias commented:
@chantra Did you get a chance to have a look at this? Thanks!

chantra commented:
Totally forgot sorry. I will check it later today.

chantra commented on cgroupstats.py:
```diff
@@ -77,13 +79,22 @@ class CgroupstatsClient(object):
+            os.close(fd)
+
+    def get_load_stats(self, path):
+        with self.open_fd(path) as fd:
```
You don't need to re-implement the context manager here. I believe:
```
with open(path, 'r') as f:
  data = CgroupstatsCmdAttrList(cgroupstats_cmd_attr_fd=f.fileno())
  ...
  ...
```

will do it just fine.

alexdias replied:
@chantra I re-implemented the context manager because using just `open` didn't work for directories (and cgroup hierarchies are under directories). Here's an example of what I get:
```
Traceback (most recent call last):
  File "cgroupstats_dump.py", line 30, in <module>
    main(sys.argv)
  File "cgroupstats_dump.py", line 25, in main
    stats = c.get_cgroup_stats(args.path)
  File "/home/alex/gnlpy2/cgroupstats.py", line 90, in get_cgroup_stats
    with open(path, 'r') as f:
IOError: [Errno 21] Is a directory: '/sys/fs/cgroup/cpu'
```

chantra replied:
@alexdias good point!

chantra commented on tests/cgroupstats_tests.py:
```diff
@@ -0,0 +1,44 @@
+        except IOError:
+            self.cgroup_available = False
+
+    def test_get_cgroup_stats(self):
```
I am tempted to think that the systems where those tests are going to run will support cgroups.
I don't think we should prevent the test from running if cgroups are not available.
Currently, I believe they are on travis, which is where the test are going to run and signal any issue.
Individual contributors can just not run this test suite if it is not supported on their setup.

Worst case, we could use `expectedFailure` decorator
https://docs.python.org/2/library/unittest.html#unittest.expectedFailure

But I rather have a test fail because of environment change than ignoring the test altogether without properly signaling it.

alexdias replied:
@chantra Yeah that makes sense. I'll get rid of that try/except on the setUp. However we can still run into the case in which the cgroup cpu subsystem is not available (it's not on the supplied Vagrant test box, for instance). If that is the case do you think we should skip the test or use the `exceptedFailure` decorator? If we were to use `skipIf` decorator (well, using our own decorator that would pass in the value of a variable such as `cpu_cgroup_present`) we can pass a reason, which I assume will be part of the test run output.

chantra replied:
Yeah, skip if sounds great! Thanks.

vmauge commented:
@alexdias I cannot see the CLA request in our internal system. Can you confirm how Oyster sent the request ?

vmauge commented:
@alexdias and as I wrote my previous comment our system finally picked-up the request so CLA is good.

alexdias commented:
@chantra I just updated the test, how does it look?

I'll squash everything to 1 commit once you give the go-ahead :)

chantra commented:
It is all good @alexdias 

The only nit, which has nothing technical, is around the file headers. When creating new files, the standard header should be used, typically the same header than other files in the directory have.
CONTRIBUTORS.md and git logs are used to keep track of contributions. Please add your name (and company if needed) to the CONTRIBUTORS.md file then squash everything.
Thanks

alexdias commented:
@chantra Just updated CONTRIBUTORS.md, the headers, and squashed everything to one commit :)

chantra commented:
Thanks a lot @alexdias 

I am waiting for the build to kick/pass and I will merge it.

Thanks again for ironing this patch out!

alexdias commented:
@chantra No worries, thank you for the review!

PS: If you plan on uploading this to PyPI, could you ping me when it's done? Just so I could update our dependencies here :)

chantra commented:
@alexdias  I will try to push it in the coming days and update this PR.

chantra commented:
@alexdias v0.1.2 was pushed to pypi!

alexdias commented:
@chantra Thank you!
